server/uploading.go

```diff
@@ -77,29 +86,37 @@ func (ut *uploadTracker) uploadStarting(r *put.Request) {
 }
 
 func (ut *uploadTracker) createAndSendSlackMsg() {
+	ut.debounceTracker.Lock()
+	defer ut.debounceTracker.Unlock()
+
 	suffix := ""
 	if len(ut.uploading) != 1 {
 		suffix = "s"
 	}
 
-	msg := fmt.Sprintf("%d client%s uploading", len(ut.uploading), suffix)
+	ut.debounceTracker.sendSlackMsg(
+		fmt.Sprintf("%d client%s uploading", len(ut.uploading), suffix))
+}
+
+func (dt *debounceTracker) sendSlackMsg(msg string) {
```

Comment:
Based on the real-world behaviour seen, I think we need to change the desired behaviour of the system to "send a message about the highest number seen in the debounce period, unless the previous debounce period had the same number as its highest". You could implement this by having newDebounceTracker() take a fmt.Sprintf with a single %d in it (eg. "%d uploading clients"), and then this method would take an int as an arg and remember the last number it emitted instead of the last msg.

There probably needs to be some kind of special handling for the last message being 0, us not emitting 0 because there was a higher number in the debounce period, but then there being no send msg requests in the following period: we need to emit that final 0 message.

Reply:
And it probably makes sense for this debounce functionality to be implemented in the slack package for general use by anything wanting to send "max number only message in debounce period" messages using slack.